File: packages/python-sdk/openoracle/adapters/chainlink_adapter.py
"""
Chainlink Oracle Adapter for OpenOracle Protocol

Provides integration with Chainlink's decentralized oracle network
for price feeds and other data sources.
"""

import aiohttp
import asyncio
from typing import Dict, List, Any
from .base_adapter import BaseOracleAdapter, OracleRequest, DataType

class ChainlinkAdapter(BaseOracleAdapter):
    """
    Chainlink oracle adapter for price feeds and market data.
    
    Usage:
        adapter = ChainlinkAdapter({
            'api_key': 'your-chainlink-api-key',
            'base_url': 'https://api.chain.link/v1'
        })
        
        response = await adapter.query(OracleRequest(
            query='BTC/USD',
            data_type=DataType.PRICE
        ))
    """
    
    def _get_adapter_name(self) -> str:
        return "chainlink"
    
    def _get_adapter_version(self) -> str:
        return "1.0.0"
    
    def _get_supported_data_types(self) -> List[DataType]:
        return [DataType.PRICE, DataType.WEATHER, DataType.SPORTS]
    
    async def _execute_query(self, request: OracleRequest) -> Any:
        """Execute Chainlink oracle query"""
        
        if request.data_type == DataType.PRICE:
            return await self._get_price_feed(request.query)
        elif request.data_type == DataType.WEATHER:
            return await self._get_weather_data(request.query)
        elif request.data_type == DataType.SPORTS:
            return await self._get_sports_data(request.query)
        else:
            raise ValueError(f"Unsupported data type: {request.data_type}")
    
    async def _get_price_feed(self, pair: str) -> Dict[str, Any]:
        """Get price feed from Chainlink"""
        base_url = self.config.get('base_url', 'https://api.chain.link/v1')
        api_key = self.config.get('api_key')
        
        headers = {}
        if api_key:
            headers['Authorization'] = f'Bearer {api_key}'
        
        async with aiohttp.ClientSession() as session:
            url = f"{base_url}/price/{pair}"
            
            async with session.get(url, headers=headers) as response:
                response.raise_for_status()
                data = await response.json()
                
                return {
                    'price': float(data['price']),
                    'pair': pair,
                    'decimals': data.get('decimals', 8),
                    'updated_at': data.get('updatedAt'),
                    'round_id': data.get('roundId')
                }
    
    async def _get_weather_data(self, location: str) -> Dict[str, Any]:
        """Get weather data from Chainlink (example)"""
        # Implement weather data fetching logic
        return {
            'location': location,
            'temperature': 22.5,
            'humidity': 65,
            'description': 'Partly cloudy'
        }
    
    async def _get_sports_data(self, event: str) -> Dict[str, Any]:
        """Get sports data from Chainlink (example)"""
        # Implement sports data fetching logic  
        return {
            'event': event,
            'status': 'completed',
            'score': {'home': 2, 'away': 1}
        }
    
    def _calculate_confidence(self, data: Any, request: OracleRequest) -> float:
        """Calculate confidence based on Chainlink data freshness"""
        if not data:
            return 0.0
        
        # Higher confidence for recent price updates
        if 'updated_at' in data:
            import time
            age_seconds = time.time() - data['updated_at']
            if age_seconds < 60:  # Less than 1 minute old
                return 0.95
            elif age_seconds < 300:  # Less than 5 minutes old
                return 0.85
            else:
                return 0.75
        
        return 0.8  # Default confidence
    
    def _calculate_cost(self, request: OracleRequest) -> float:
        """Calculate cost for Chainlink query"""
        # Chainlink pricing model (example)
        if request.data_type == DataType.PRICE:
            return 0.001  # $0.001 per price query
        return 0.005  # $0.005 for other data types
    
    async def _health_check_query(self) -> Any:
        """Chainlink health check"""
        return await self._get_price_feed("ETH/USD")

File: packages/node-sdk/src/adapters/ChainlinkAdapter.ts
/**
 * Chainlink Oracle Adapter for OpenOracle Protocol (Node.js/TypeScript)
 * 
 * Provides integration with Chainlink's decentralized oracle network
 * for price feeds and other data sources.
 */

import axios, { AxiosResponse } from 'axios'
import { BaseOracleAdapter, OracleRequest, DataType, AdapterConfig } from './BaseOracleAdapter'

export interface ChainlinkConfig extends AdapterConfig {
  apiKey?: string
  baseUrl?: string
  network?: 'mainnet' | 'testnet'
}

/**
 * Chainlink oracle adapter for price feeds and market data.
 * 
 * Usage:
 * ```typescript
 * const adapter = new ChainlinkAdapter({
 *   apiKey: 'your-chainlink-api-key',
 *   baseUrl: 'https://api.chain.link/v1'
 * })
 * 
 * const response = await adapter.query({
 *   query: 'BTC/USD',
 *   dataType: DataType.PRICE
 * })
 * ```
 */
export class ChainlinkAdapter extends BaseOracleAdapter {
  
  constructor(config: ChainlinkConfig) {
    super(config)
  }

  protected getAdapterName(): string {
    return 'chainlink'
  }

  protected getAdapterVersion(): string {
    return '1.0.0'
  }

  protected getSupportedDataTypes(): DataType[] {
    return [DataType.PRICE, DataType.WEATHER, DataType.SPORTS]
  }

  protected async executeQuery(request: OracleRequest): Promise<any> {
    switch (request.dataType) {
      case DataType.PRICE:
        return this.getPriceFeed(request.query)
      case DataType.WEATHER:
        return this.getWeatherData(request.query)
      case DataType.SPORTS:
        return this.getSportsData(request.query)
      default:
        throw new Error(`Unsupported data type: ${request.dataType}`)
    }
  }

  /**
   * Get price feed from Chainlink
   */
  private async getPriceFeed(pair: string): Promise<any> {
    const baseUrl = this.config.baseUrl || 'https://api.chain.link/v1'
    const headers: Record<string, string> = {
      'Content-Type': 'application/json'
    }

    if (this.config.apiKey) {
      headers['Authorization'] = `Bearer ${this.config.apiKey}`
    }

    try {
      const response: AxiosResponse = await axios.get(
        `${baseUrl}/price/${pair}`,
        { headers }
      )

      return {
        price: parseFloat(response.data.price),
        pair,
        decimals: response.data.decimals || 8,
        updatedAt: response.data.updatedAt,
        roundId: response.data.roundId,
        source: 'chainlink'
      }
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(`Chainlink API error: ${error.response?.status} ${error.response?.statusText}`)
      }
      throw error
    }
  }

  /**
   * Get weather data from Chainlink (example implementation)
   */
  private async getWeatherData(location: string): Promise<any> {
    // Example implementation - replace with actual Chainlink weather oracle
    const baseUrl = this.config.baseUrl || 'https://api.chain.link/v1'
    
    // Mock implementation for demonstration
    return {
      location,
      temperature: 22.5,
      humidity: 65,
      description: 'Partly cloudy',
      source: 'chainlink-weather',
      timestamp: Date.now()
    }
  }

  /**
   * Get sports data from Chainlink (example implementation)
   */
  private async getSportsData(event: string): Promise<any> {
    // Example implementation - replace with actual Chainlink sports oracle
    return {
      event,
      status: 'completed',
      score: { home: 2, away: 1 },
      source: 'chainlink-sports',
      timestamp: Date.now()
    }
  }

  protected calculateConfidence(data: any, request: OracleRequest): number {
    if (!data) return 0.0

    // Higher confidence for recent price updates
    if (data.updatedAt) {
      const ageSeconds = (Date.now() - data.updatedAt) / 1000
      if (ageSeconds < 60) return 0.95      // Less than 1 minute old
      if (ageSeconds < 300) return 0.85     // Less than 5 minutes old
      return 0.75                           // Older data
    }

    return 0.8 // Default confidence
  }

  protected calculateCost(request: OracleRequest): number {
    // Chainlink pricing model (example)
    const pricing: Record<DataType, number> = {
      [DataType.PRICE]: 0.001,    // $0.001 per price query
      [DataType.WEATHER]: 0.005,  // $0.005 per weather query
      [DataType.SPORTS]: 0.005,   // $0.005 per sports query
      [DataType.CUSTOM]: 0.01,
      [DataType.NEWS]: 0.01,
      [DataType.SOCIAL]: 0.01
    }

    return pricing[request.dataType] || 0.01
  }

  protected getResponseMetadata(request: OracleRequest): Record<string, any> {
    return {
      provider: 'chainlink',
      network: this.config.network || 'mainnet',
      version: this.version
    }
  }

  protected async healthCheckQuery(): Promise<any> {
    // Simple health check using ETH/USD price feed
    return this.getPriceFeed('ETH/USD')
  }
}